### PR TITLE
Resurrect a database unit test

### PR DESCRIPTION
**Pull request by jonludlam**
Signed-off-by: Jon Ludlam <jonathan.ludlam@citrix.com>

### CODE REVIEW

File: ocaml/database/jbuild [hunk 1]
```diff
@@ -71,4 +72,41 @@ let () = Printf.ksprintf Jbuild_plugin.V1.send {|
...
+    ))
+  (libraries (
+    xapi-database
+  ))
+ )
+)
+
+(executable
+ ((name database_server_main)
+  (public_name database_server)
```
Comment (edwintorok) [inline ocaml/database/jbuild]: If you set a public name it'll get installed, is that what you intended?

File: ocaml/database/schema.ml [hunk 2]
```diff
@@ -82,7 +82,11 @@ module Table = struct
     columns: Column.t list;
     persistent: bool;
   } [@@deriving sexp]
-  let find name t = List.find (fun col -> col.Column.name = name) t.columns
+  let find name t =
+    try
+      List.find (fun col -> col.Column.name = name) t.columns
+    with Not_found ->
```
Comment (edwintorok) [inline ocaml/database/schema.ml]: Thanks it is always annoying when you get `Not_found` in the logs not knowing what was not found...

File: ocaml/database/database_server_main.ml [hunk 3]
```diff
@@ -1,3 +1,6 @@
+open Xapi_stdext_threads
+open Xapi_stdext_unix
+open Xapi_stdext_threads
```
Comment (mseri) [inline ocaml/database/database_server_main.ml]: This is already opened above

File: ocaml/database/database_test.ml [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+      let real_ref = if Client.is_valid_ref t key then key else Client.db_get_by_uuid t tblname key in
+      let real_name_label =
+        try Some (Client.read_field t tblname name_label real_ref)
+        with _ -> None in
+      if name_label' <> real_name_label
+      then failwith (Printf.sprintf "check_ref_index %s key %s: ref_index name_label = %s; db has %s" tblname key (Opt.default "None" name_label') (Opt.default "None" real_name_label))
+
+
+open Xapi_stdext_pervasives
+open Xapi_stdext_monadic
```
Comment (mseri) [inline ocaml/database/database_test.ml]: This is opened above

File: ocaml/database/schema.ml (hunk 2 above)
Comment (mseri) [reply ocaml/database/schema.ml]: Indeed!

File: ocaml/database/test_schemas.ml [hunk 5]
```diff
@@ -74,12 +82,13 @@ let schema =
 
   let vm_table = {
     Schema.Table.name = "VM";
-    columns = [ _ref; uuid; name_label; vbds; pp; name_description; tags; other_config ];
+    columns = [ _ref; uuid; name_label; vbds; pp;
+    name_description; tags; other_config ];
```
Comment (mseri) [inline ocaml/database/test_schemas.ml]: Did you really mean to split them like this?

File: ocaml/database/jbuild (hunk 1 above)
Comment (jonludlam) [reply ocaml/database/jbuild]: Ah yes, copy-pasta mistake - thanks!

File: ocaml/database/database_test.ml [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+      set_field "foo" "foo:1" "bars" (add_to_set "bar:1" (Set []))
+    in
+    (* check that 'bar.foos' includes 'foo' *)
+    let bar_1 = Table.find "bar:1" (TableSet.find "bar" (Database.tableset db)) in
+    let bar_foos = Row.find "foos" bar_1 in
+    if bar_foos <> Set ["foo:1"]
+    then failwith (Printf.sprintf "check_many_to_many: bar(bar:1).foos expected ('foo:1') got %s" (Schema.Value.marshal bar_foos));
+
+    (* set foo.bars to [] *)
+    (*		let foo_1 = Table.find "foo:1" (TableSet.find "foo" (Database.tableset db)) in*)
```
Comment (mseri) [inline ocaml/database/database_test.ml]: Why you left this line commented out?

File: ocaml/database/database_test.ml [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+    let deleted = get_deleted db g9 in
+    if (List.mem vbd deleted)
+    then (Printf.printf "Pass\n")
+    else (Printf.printf "Fail\n"; failwith "Event problem");
+    dump db g9;
+    ignore(g10);
+
+
+
+    ()
```
Comment (mseri) [inline ocaml/database/database_test.ml]: The choice of spacing and newlines are "interesting" in some part of this code :stuck_out_tongue_winking_eye:

File: ocaml/database/database_test.ml [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+    let where_name_label =
+      { Db_cache_types.table = "VM"; return = name_label; where_field="uuid"; where_value = valid_uuid } in
+    let xs = Client.read_field_where t where_name_label in
+    if not (List.mem name xs)
+    then failwith "read_field_where <valid table> <valid return> <valid field> <valid value>";
+    test_invalid_where_record "read_field_where" (Client.read_field_where t);
+
+(*    let xs = Client.read_set_ref t where_name_label in
+    if not (List.mem name xs)
+    then failwith "read_set_ref <valid table> <valid return> <valid field> <valid value>";
```
Comment (mseri) [inline ocaml/database/database_test.ml]: Why you disabled this subtest?

File: ocaml/database/database_test.ml (hunk 7 above)
Comment (jonludlam) [reply ocaml/database/database_test.ml]: Ah, I've forgotten to ocp-indent it.

**Pull request merged**